core/http-auth/src/main/java/software/amazon/awssdk/http/auth/internal/DefaultAwsV4HttpSigner.java

```diff
@@ -15,27 +15,228 @@
 import software.amazon.awssdk.http.auth.spi.AsyncSignedHttpRequest;
+import software.amazon.awssdk.http.auth.spi.HttpSignRequest;
 import software.amazon.awssdk.http.auth.spi.SyncHttpSignRequest;
 import software.amazon.awssdk.http.auth.spi.SyncSignedHttpRequest;
 import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.identity.spi.AwsSessionCredentialsIdentity;
+import software.amazon.awssdk.utils.BinaryUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.StringUtils;
+import software.amazon.awssdk.utils.http.SdkHttpUtils;
 
 /**
  * A default implementation of {@link AwsV4HttpSigner}.
  */
 @SdkInternalApi
 public class DefaultAwsV4HttpSigner implements AwsV4HttpSigner {
 
+    public static final String UNSIGNED_PAYLOAD = "UNSIGNED-PAYLOAD";
+
+    private static final Logger LOG = Logger.loggerFor(DefaultAwsV4HttpSigner.class);
+
     @Override
     public SyncSignedHttpRequest sign(SyncHttpSignRequest<? extends AwsCredentialsIdentity> request) {
-        throw new UnsupportedOperationException();
+        // anonymous credentials, don't sign
+        if (CredentialUtils.isAnonymous(request.identity())) {
+            return SyncSignedHttpRequest.builder()
+                .request(request.request())
+                .payload(request.payload().orElse(null))
+                .build();
+        }
+
+        SdkHttpRequest signedRequest = doSign(request).build();
+        return SyncSignedHttpRequest.builder()
+            .request(signedRequest)
+            .payload(request.payload().orElse(null))
+            .build();
+    }
+
+    private SdkHttpRequest.Builder doSign(SyncHttpSignRequest<? extends AwsCredentialsIdentity> request) {
+        SdkChecksum sdkChecksum = createSdkChecksumFromRequest(request.request(), request.property(CHECKSUM_HEADER_NAME),
+            request.property(CHECKSUM_ALGORITHM));
+        String contentHash = calculateContentHash(request.payload().orElse(null), sdkChecksum);
+        return doSign(request, new ContentChecksum(contentHash, sdkChecksum));
+    }
+
+    private SdkHttpRequest.Builder doSign(HttpSignRequest<?, ? extends AwsCredentialsIdentity> request,
+                                          ContentChecksum contentChecksum) {
+        SdkHttpRequest.Builder requestBuilder = request.request().toBuilder();
+        Boolean doubleUrlEncode = request.property(DOUBLE_URL_ENCODE);
+        Boolean normalizePath = request.property(NORMALIZE_PATH);
+        String checksumHeaderName = request.property(CHECKSUM_HEADER_NAME);
+        Instant requestSigningInstant = request.property(REQUEST_SIGNING_INSTANT);
+        String formattedRequestSigningDate = formatDateStamp(requestSigningInstant);
+        String formattedRequestSigningDateTime = formatTimestamp(requestSigningInstant);
+        String regionName = request.property(REGION_NAME);
+        String serviceSigningName = request.property(SERVICE_SIGNING_NAME);
```

Comment:
we should validate required signerProperties upfront and throw instead of NPE other unexpected errors. Would properties like doubleUrlEncode have default values when not provided? The required/default aspects of signer properties should be documented in the interface.

Reply:
Looking at `Aws4SignerParams`, it appears there are a couple of defaults, yes. I'll add them. If we add the aspects to the interface, it should probably be in terms of "MUST/SHOULD/etc."